llpc/context/llpcCompiler.cpp

```diff
@@ -728,6 +728,246 @@ template <typename T> static FormattedBytes formatBytesLittleEndian(ArrayRef<T>
   return format_bytes(bytes, /* FirstByteOffset = */ None, /* NumPerLine = */ 16, /* ByteGroupSize = */ 1);
 }
 
+// Build shader from the specified info.
+//
+// @param [in]  pipelineInfo : Info to build this shader module
+// @param [out] pipelineOut : Output of building this shader module
+// @param [in]  stage : Shader stage of needing to compile
+// @param pPipelineDumpFile : Handle of pipeline dump file
+//
+// @returns : Result::Success if successful. Other return codes indicate failure.
+Result Compiler::buildGraphicsShaderStage(const GraphicsPipelineBuildInfo *pipelineInfo,
+                                          GraphicsPipelineBuildOut *pipelineOut, UnlinkedShaderStage stage,
+                                          void *pPipelineDumpFile) {
+  Result result = Result::Success;
+
+  BinaryData elfBin = {};
+  // clang-format off
+  SmallVector<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfo = {
+    &pipelineInfo->vs,
+    &pipelineInfo->tcs,
+    &pipelineInfo->tes,
+    &pipelineInfo->gs,
+    &pipelineInfo->fs,
+  };
+  // clang-format on
+
+  MetroHash::Hash cacheHash = {};
+  MetroHash::Hash pipelineHash = {};
+  cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+  pipelineHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+
+  if (pPipelineDumpFile) {
+    std::stringstream strStream;
+    strStream << ";Compiler Options: ";
+    for (auto &option : m_options)
+      strStream << option << " ";
+    std::string extraInfo = strStream.str();
+    PipelineDumper::DumpPipelineExtraInfo(reinterpret_cast<PipelineDumpFile *>(pPipelineDumpFile), &extraInfo);
+  }
+
+  Optional<CacheAccessor> cacheAccessor;
+  ElfPackage candidateElf;
+
+  if (!cacheAccessor || !cacheAccessor->isInCache()) {
+    LLPC_OUTS("Cache miss for graphics pipeline.\n");
+
+    // Compile
+    GraphicsContext graphicsContext(m_gfxIp, pipelineInfo, &pipelineHash, &cacheHash);
+    Context *context = acquireContext();
+    context->attachPipelineContext(&graphicsContext);
+
+    result = buildUnlinkedShaderInternal(context, shaderInfo, stage, candidateElf, pipelineOut->stageCacheAccesses);
+    releaseContext(context);
+
+    if (result == Result::Success) {
+      elfBin.codeSize = candidateElf.size();
+      elfBin.pCode = candidateElf.data();
+    }
+
+    if (cacheAccessor && pipelineOut->pipelineCacheAccess == CacheAccessInfo::CacheNotChecked)
+      pipelineOut->pipelineCacheAccess = CacheAccessInfo::CacheMiss;
+  } else {
+    LLPC_OUTS("Cache hit for graphics pipeline.\n");
+    elfBin = cacheAccessor->getElfFromCache();
+    if (cacheAccessor->isInCache()) {
+      pipelineOut->pipelineCacheAccess =
+          cacheAccessor->hitInternalCache() ? CacheAccessInfo::InternalCacheHit : CacheAccessInfo::CacheHit;
+    }
+  }
+
+  if (result == Result::Success) {
+    void *allocBuf = nullptr;
```

Comment:
`allocBuf` can be defined inside the next if.

Reply:
Fixed, Thanks.